rclcpp/include/rclcpp/scope_exit.hpp

```diff
@@ -1,52 +0,0 @@
-// Copyright 2015 Open Source Robotics Foundation, Inc.
-//
-// Licensed under the Apache License, Version 2.0 (the "License");
-// you may not use this file except in compliance with the License.
-// You may obtain a copy of the License at
-//
-//     http://www.apache.org/licenses/LICENSE-2.0
-//
-// Unless required by applicable law or agreed to in writing, software
-// distributed under the License is distributed on an "AS IS" BASIS,
-// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-// See the License for the specific language governing permissions and
-// limitations under the License.
-
-// Based on: http://the-witness.net/news/2012/11/scopeexit-in-c11/
-// But I changed the lambda to include by reference rather than value, see:
-// http://the-witness.net/news/2012/11/scopeexit-in-c11/comment-page-1/#comment-86873
-
-#ifndef RCLCPP__SCOPE_EXIT_HPP_
-#define RCLCPP__SCOPE_EXIT_HPP_
-
-#include <functional>
-
-#include "rclcpp/macros.hpp"
-
-namespace rclcpp
-{
-
-template<typename Callable>
-struct ScopeExit
```

Comment:
I think we should deprecate this header before removing it, if it is technically possible. We can, and have, removed things without first deprecating them (but most often out of technical necessity), and think this code is not widely used, but it is in the public API and we should at least try to do a deprecation first.

We can put a `#warning` in this header and possibly even just alias this class to the rcpputils version, assuming the APIs are the same. (also for the macro)

Reply:
done in 68adbafe7566d76729ce3f405648e59ee8b68caf

To keep it simple I just kept the file and added a warning.

Reply:
I found some files in system_tests indirectly including this header, so this PR will break that (and other instances of it), but in any case I think we're supposed to "include what we use." We'll see with CI.

Reply:
Ok, well another option is that we can wrap the `#warrning` in a conditional, and when including it from within rclcpp we can disable the warning, that way it only applies when external users include it directly.

Or we could just deprecate the `struct ScopeExit`, so people stop using it. I guess they might still be including it even if they stop using it, but then I feel less bad about the header disappearing in the next version.

Reply:
> in any case I think we're supposed to "include what we use."

I guess you mean, if we change the rclcpp headers so this header is not implicitly included, and then that breaks someone that's their fault for not "including what they use", which I'm also ok with.

Reply:
Do you want to do anything here or merge it as-is?

Reply:
> I guess you mean, if we change the rclcpp headers so this header is not implicitly included, and then that breaks someone that's their fault for not "including what they use", which I'm also ok with.

yes exactly.

> Do you want to do anything here or merge it as-is?

Since you're okay with breaking the transitive include, I'd just stick with this for simplicity.